### PR TITLE
Pass version information to environment variables for Sentry

### LINKED ISSUE #547: Changing to create entities on receipt: reprocessing submissions that resulted in an entity.error
Right now, when you configure an entity list to create entities on submission receipt, you have the option to process pending submissions. Submissions that have already created or updated an entity are not reprocessed. However, I'm noticing that submissions that were processed, but resulted in a processing error (an `entity.error` event) _are_ reprocessed. I don't think that matters much for submissions that create entities, but I think it has more potential for unexpected outcomes for submissions that update entities.

As noted in getodk/central-backend#1044, it's not known before a submission is processed for entities whether it will have an effect on entities at all, and if it will, whether it's to create or update an entity. Processing pending submissions may process submissions that update entities. I agree with the reasoning in the PR that that's not an issue. However, I think it's more of an issue if such a submission is *re*processed, and that reprocessing is successful even when the first attempt at processing failed.

#### Steps to reproduce the problem

Set things up:

- Upload a form that creates an entity list. Publish it.
- Configure the entity list to create entities on submission approval.

Set up a problematic submission that tries to create an entity:

- Create a submission that fails to create an entity (e.g., the UUID is invalid, or the label is blank).
- Approve the submission.
- Observe that the submission results in an `entity.error`.
- Mark the submission as "has issues".

Set up up a problematic submission that tries to update an entity:

- Create a second submission that would create an entity on approval. Don't approve it yet.
- Create a third submission to update the entity that the second submission would create.
- The entity doesn't exist yet, so the third submission should quickly result in an `entity.error`.

Observe that problematic submissions are reprocessed:

- Change the entity list to create entities on submission receipt. Select yes for processing pending submissions.
- Observe that the submission that failed to create an entity has been reprocessed. There are now two `entity.error`s.
- Observe that the submission that failed to update an entity has been reprocessed. The second update attempt was successful.

#### URL of the page

- Project: https://staging.getodk.cloud/#/projects/50
- Submission that failed to create an entity: https://staging.getodk.cloud/#/projects/50/forms/trees_registration_demo/submissions/dc9c80a2-7cf3-4f57-98e9-633f9529aceb
- Submission that initially failed to update an entity: https://staging.getodk.cloud/#/projects/50/forms/trees_registration_demo/submissions/9f9e9193-a96c-4815-a147-1a9c94bc4bef

#### Expected behavior

I'm not sure what the right behavior is. Maybe it's actually nice that if an update failed because an entity didn't exist, the update will be reprocessed. However, I also think that makes it harder to reason about the effects of a change to an entity list's workflow (changing it to create entities on receipt and having it process pending submissions).

Two ideas:

1. Reprocess submissions with errors, but mention that behavior in Frontend.
2. Don't reprocess submissions with errors. I think there's a case to be made that the only thing that should fix a submission with an `entity.error` is a submission edit.

Part of me thinks that it'd be nice to completely separate a change to an entity list workflow from entity updates. In other words, changing an entity list to create entities on submission receipt and processing pending submissions may _create_ entities, but it should never update them. However, I don't know if that separation would play well with the corner case of a submission that either updates or creates entities (`create="1" update="1"`).

#### Central version shown in version.txt

```
versions:
15e519a04eefc41049c30d53b6779682f7890764 (v2023.4.0-3-g15e519a)
+f4a1367ed71b2b79306247c30080d42d988418e4 client (v2023.4.0-27-gf4a1367e)
+d444524255f2ab0eadacb82e3436dcec0ae3d198 server (v2023.4.0-37-gd4445242)
```

### PR DESCRIPTION
**Pull request by alxndrsn**
Closes https://github.com/getodk/central-backend/issues/547?

Related: #316

### CODE REVIEW

**Comment by yanokwa** [comment]
We have some version information at https://sandbox.getodk.cloud/version.txt. Would we be able to use that rather than generate something new?

I think it'd be nice to have both backend and frontend versions. Not sure how best to get all that into Sentry.

**Comment by alxndrsn** [comment]
version.txt is in the nginx container, whereas the backend code is running in the service container, so i'm not sure they could use the same file.

Restrictions on content:

> The release name cannot:
>
> * contain newlines, tabulator characters, forward slashes(/) or back slashes(\)
> * be (in their entirety) period (.), double period (..), or space ( )
> * exceed 200 characters
>
> -- _https://docs.sentry.io/platforms/javascript/configuration/releases/#bind-the-version_

An example `versions.txt` with linebreaks removed:

```
versions: d774e91d53a0ed1ce9c46109aadffa8add2c8080 (v1.5.2) 3078965dc8ad86b34944476d0dd84475487d44ab client (v1.5.2) 12568c504151c919b11c9962dd3ef866cebbd0e7 server (v1.5.1)
```
length: 174 characters

@yanokwa would that be suitable?

**Comment by alxndrsn** [comment]
@yanokwa `git describe --tags --dirty` gets us a usable git tree-ish while being a bit more readable; would the following format be suitable:

```
central:{central-version}; backend:{backend-version}; frontend:{frontend-version}
```

e.g.
```
central:v1.5.2-5-g43d3897-dirty; backend:v1.5.1-73-g201b2216; frontend:v1.5.1
```

?

This would also leave room for enketo-express version.

However, both approaches, and the current `version.txt`, have the disadvantage of not necessarily being accurate - if any other service is changed without rebuilding the `central_service` or `central_nginx` containers, then the running versions may have changed independent of what's included in the `central_service` container.

**Comment by yanokwa** [comment]
Maybe we match the version.txt a little bit as far as the keys (server/client vs backend/frontend)? We can even make it shorter and drop first key and the semi-colon.
`{central-version} server:{backend-version} client:{frontend-version}`.

Thanks for noting the accuracy limitation. I'm OK with things not being super accurate. Our docs all have people rebuild everything. Also, this info we collect just helps us narrow down where a problem might be.

**Comment by matthew-white** [comment]
Just wanted to chime in here that I like the idea of specifying `--dirty` for `git describe`. I don't think we do something similar for version.txt: version.txt doesn't indicate whether the working tree has local modification. But I think that's something that'd be useful to know in Sentry. (And maybe we should indicate that in version.txt as well.)

**Comment by matthew-white** [comment]
I'm now wondering whether we should try to use different tags for `central`/Backend/Frontend instead of combining the three into a single release tag. (Probably that would mean using the release tag for the Backend version and [other tags](https://docs.sentry.io/platforms/javascript/enriching-events/tags/) for `central` and Frontend.)

The reason I'm thinking that is that I think we may want to filter out events that don't match a known version of Central. We get a fair number of Sentry events from ODK forks. (Recent examples: an [error](https://sentry.io/organizations/getodk/issues/3544185221) from an endpoint that doesn't exist in our codebase, an [error](https://sentry.io/organizations/getodk/issues/3554400571) from a file that doesn't exist in our codebase, another [error](https://sentry.io/organizations/getodk/issues/3561525016) from a file that doesn't exist.) However, I think that if we combine `central`, Backend, and Frontend into a single tag, it'll be harder to filter for known versions of Central, because the intersection of the three versions can result in a sizable list:

- Sometimes the `central` version will match a release tag: that's the case with ODK Cloud. Outside of ODK Cloud, it's more common for the version to be one or commits ahead of the latest release tag, since we update the [news](https://github.com/getodk/central/blob/master/docs/news.html) after publishing a release, and users deploy from the `master` branch. Anytime we merge a no-op commit (for example, updating the readme), that'll be a different version that could show up in Sentry. Sometimes the `central` working directory will be clean, but in other cases (including with ODK Cloud), it won't be. If we use `git describe --dirty` (as I think we should), that difference will show up in Sentry. Altogether, we can expect to see more valid "versions" of `central` in Sentry than there are actual releases.
- Frontend is simpler: the Frontend version will always match a release tag. However, ODK Cloud deployments modify [`src/config.js`](https://github.com/getodk/central-frontend/blob/master/src/config.js), so the working directory won't be clean.
- Backend is simplest: the Backend version will always match a release tag. If the version doesn't match a release tag, or if the working directory isn't clean, that points to a fork.

Instead of trying to filter on the combination of these versions, I think we'll probably want to start by filtering solely on the Backend version. We can then additionally filter on the other two versions as needed. To do that, I think we should use three separate tags. I think we could make all three normal tags, or we could use the release tag for the Backend version and normal tags for the other two.

**Comment by matthew-white** [comment]
Does the approach above sound reasonable? I'd be happy to run with that idea and add to this PR if there's an interest in taking that approach.

**Comment by alxndrsn** [comment]
> Instead of trying to filter on the combination of these versions, I think we'd probably want to start by filtering solely on the Backend version. We could then additionally filter on the other two versions as needed. To enable that, I think we should use three separate tags. I think we could make all three normal tags, or we could use the release tag for the Backend version and normal tags for the other two.

Sounds good to me :+1:

**Comment by matthew-white** [comment]
I've merged in the `release` branch in order to resolve a conflict with #318. I've also pushed a commit to give a sense of the approach I have in mind. I've left some comments about questions I have. @alxndrsn, would you mind taking a look at my changes? Once we finalize these changes, I'll work on a corresponding PR to Backend.

File: service.dockerfile [hunk 1]
```diff
@@ -10,6 +20,7 @@ RUN echo "deb http://apt.postgresql.org/pub/repos/apt/ stretch-pgdg main" | tee
 COPY files/service/crontab /etc/cron.d/odk
 
 COPY server/package*.json ./
+COPY --from=intermediate /tmp/sentry-versions/ ./
```
Comment (alxndrsn) [inline service.dockerfile:23]: Maybe this would be more efficient if placed last?

**Comment by alxndrsn** [comment]
@matthew-white this looks good to me.  I can't approve the PR, as I'm the original author :shrug:

**Comment by matthew-white** [comment]
There was a conflict with #334, so I've pushed a commit to resolve that.

File: files/service/scripts/start-odk.sh [hunk 2]
```diff
@@ -26,6 +26,9 @@ else
 fi
 echo "using $WORKER_COUNT worker(s) based on available memory ($MEMTOT).."
 
+export SENTRY_RELEASE="$(cat sentry-versions/server)"
+export SENTRY_TAGS="{ \"version.central\": \"$(cat sentry-versions/central)\", \"version.client\": \"$(cat sentry-versions/client)\" }"
```
[the remark this answers is not in the document]
Comment (matthew-white) [reply files/service/scripts/start-odk.sh]: I ended up setting these variables earlier, before migrations are run. Other environment variables are set in the same place, so it feels like a good spot to be setting these variables.

**Review by matthew-white — APPROVED** [review]
This has been a joint effort, but I'll go ahead and give the checkmark. ✅

**Comment by alxndrsn** [comment]
> This has been a joint effort, but I'll go ahead and give the checkmark. white_check_mark

Github wouldn't allow me to approve.  Thanks for all your work on this :+1:

**Pull request merged**